### ISSUE TITLE
Fission 1.11.0 environment declarative specs are not working

### ISSUE THREAD

**Issue by abbaspw21** [fission/fission#1805]
**Fission/Kubernetes version**

<pre>
$ fission version
client:
  fission/core:
    BuildDate: "2020-09-16T15:37:59Z"
    GitCommit: 09b4ffed78ad7a7eea7ebaf47e7644843b0a0b12
    Version: 1.11.0
server:
  fission/core:
    BuildDate: "2020-09-16T13:41:28Z"
    GitCommit: f034a1c734af51cdd6dd4449d1db253a67e699c3
    Version: 1.11.0

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.1", GitCommit:"d647ddbd755faf07169599a625faf302ffc34458", GitTreeState:"clean", BuildDate:"2019-10-02T17:01:15Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.1", GitCommit:"d647ddbd755faf07169599a625faf302ffc34458", GitTreeState:"clean", BuildDate:"2019-10-02T16:51:36Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}

</pre>

**Describe the bug**
After Upgrading fission to **1.11.0 from 1.9.0** declarative specs for fission environment are not working as expected. The configurations present in specs template file for **resources**, **runtime.container** and **runtime.podspec** are not getting applied.

**To Reproduce**
To reproduce the issue create fission environment using below specs template file and then fetch the K8 resource created for the fission environment :
```
apiVersion: fission.io/v1
kind: Environment
metadata:
  creationTimestamp: null
  name: test
  namespace: fission-function
spec:
  TerminationGracePeriod: 60
  builder: {}
  keeparchive: false
  poolsize: 3
  resources:
    limits:
        cpu: 1
        memory: 128Mi
    requests:
        cpu: 10m
        memory: 16Mi
  runtime:
    functionendpointport: 0
    image: <ENV_IMAGE>
    container:
       env:
       - name: ENABLE_TRACING
         value: "true"
       - name: ENABLE_LOGGING
         value: "true"
       volumeMounts:
       - mountPath: /etc/localtime
         name: tz-config
    loadendpointpath: ""
    loadendpointport: 0
    podspec:
      volumes:
      - hostPath:
          path: /etc/localtime
          type: ""
        name: tz-config
  version: 3

```

**Expected result** 
<pre>
kubectl get environment test  -n fission-function -o yaml

apiVersion: fission.io/v1
kind: Environment
metadata:
  annotations:
    fission-name: test
    fission-uid: 3146ef85-0e0a-4cd7-baf7-152f96409086
  creationTimestamp: "2020-09-15T15:54:27Z"
  generation: 1
  name: test
  namespace: fission-function
  resourceVersion: "8888"
  selfLink: /apis/fission.io/v1/namespaces/fission-function/environments/test
  uid: 33523f0f-fb08-4802-b0db-bc05f7d30a8a
spec:
  builder: {}
  imagepullsecret: ""
  keeparchive: false
  poolsize: 3
  resources:
    limits:
      cpu: "1"
      memory: 128Mi
    requests:
      cpu: 10m
      memory: 16Mi
  runtime:
    container:
      env:
      - name: ENABLE_TRACING
        value: "true"
      - name: ENABLE_LOGGING
        value: "true"
      name: ""
      resources: {}
      volumeMounts:
      - mountPath: /etc/localtime
        name: tz-config
    image: <ENV_IMAGE>
    podspec:
      containers: null
      volumes:
      - hostPath:
          path: /etc/localtime
          type: ""
        name: tz-config
  terminationGracePeriod: 60
  version: 3
</pre>

**Actual result**
<pre>
kubectl get environment test  -n fission-function -o yaml

apiVersion: fission.io/v1
kind: Environment
metadata:
  annotations:
    fission-name: test
    fission-uid: 3146ef85-0e0a-4cd7-baf7-152f96409086
  creationTimestamp: "2020-10-03T13:14:38Z"
  generation: 1
  name: test
  namespace: fission-function
  resourceVersion: "523333"
  selfLink: /apis/fission.io/v1/namespaces/fission-function/environments/test
  uid: e028c119-20bf-413d-b11c-11e0916587e7
spec:
  builder: {}
  imagepullsecret: ""
  keeparchive: false
  poolsize: 3
  resources: {}
  runtime:
    container: {}
    image: <ENV_IMAGE>
    podspec: {}
  terminationGracePeriod: 60
  version: 3
</pre>

As you can see above the configurations defined in **resources**, **runtime.container** and **runtime.podspec** are missing in K8 resource for fission environment. The same configuration works fine in 1.9.0

**Comment by abbaspw21** [comment]
Tagging @vishal-biyani

**Comment by vishal-biyani** [comment]
Hey @abbaspw21 Thanks, will take a look at priority!

**Comment by mattroark** [comment]
Appreciate this being addressed on priority. I can validate that I'm hitting the same on 1.11.0 – `podspec` is not populated despite it being defined in the spec file.

{code}
server:
  fission/core:
    BuildDate: "2020-09-16T13:41:28Z"
    GitCommit: f034a1c734af51cdd6dd4449d1db253a67e699c3
    Version: 1.11.0
{code}

**Comment by justinfx** [comment]
Also want to push for this as priority. I want to pick up the latest Keda support and start exploring Keda kafka triggers but I also rely on the podSpec

**Comment by vishal-biyani** [comment]
Closed by https://github.com/fission/fission/pull/1818 - landing in 1.11.1 - to be released shortly.

**Comment by abbaspw21** [comment]
Hi @vishal-biyani / @therahulbhati,

Even in release 1.11.1 fission environment declarative specs (podspec) are not working

**Comment by vishal-biyani** [comment]
Hi @abbaspw21 I think the PodSpec itself is working with 1.11.1 but the resources part still not working - we spotted the CRD validation might have introduced this and might also have issues with specific versions of K8S - can you please confirm the K8S version? We will make another release today.

**Comment by abbaspw21** [comment]
Hi @vishal-biyani ,

Our current Kubernetes version is 1..16.1.

It seems that this issue occurs when we are upgrading fission from 1.11.0 to 1.11.1.
In case of fresh installation or while upgrading it from previous version it seems to be working fine. 

We are good with it as we are going to upgrade our environments from 1.9.0 to 1.11.1.

**Comment by vishal-biyani** [comment]
Thanks for confirming @abbaspw21 and glad to see it is working. We will also do a few more rounds of validation and post if we find anything more! I will keep the issue open for a few more days till we validate and then close.

**Comment by therahulbhati** [comment]
Hi @abbaspw21,

We introduced CRD validation in release 1.11.0. There was a bug in that due to which resources, runtime.container and runtime.podspec were getting pruned.

CRD validations are only applicable if you recreate fission CRs. 
If you just upgrade from 1.9.0 to 1.11.0 or 1.11.1 you won't face this issue because CRD validations are not happening, since you aren't recreating fission CRs. 
However, you would face this pruning of fields issue, if you delete old CRs and then install 1.11.0 or install 1.11.0 on a fresh cluster because here CRs are created again.
 
We tried a few things as below:
- Installed release 1.9.0 -> Upgraded to 1.11.0. Works as expected (no field is pruned)
- Installed release 1.9.0 -> Upgraded to 1.11.1. Works as expected (no field is pruned)
- Installed release 1.10.0 -> Upgraded to 1.11.0. Works as expected (no field is pruned)
- Installed release 1.10.0 -> Upgraded to 1.11.1. Works as expected (no field is pruned)
- Install release 1.11.0 on fresh cluster. All mentioned field in the issue gets pruned.
- Install release 1.11.1 on fresh cluster. Only resources field gets pruned.

Can you confirm one thing, did you recreated CRs while upgrading or something like that while upgrading from 1.9.0 to 1.11.0?

**Comment by abbaspw21** [comment]
Hi @therahulbhati ,

Yes, we were recreating the CRs after upgrading fission from 1.9.0 to 1.11.0.

As already mentioned by you in above comment, we have also noticed that resources field is getting pruned in case of fresh installation or upgrade if CRs are recreated.

**Comment by abbaspw21** [comment]
Hi @vishal-biyani/ @therahulbhati ,

We have tested declarative specs in release 1.11.2 and it seems to be working fine.

Thank you !!!

**Comment by airhorns** [comment]
I observed this behaviour too and 1.11.2 seems to have fixed it for me!

I installed fission with helm, and it seems like `helm upgrade` didn't update the CRDs properly, so even though I started using the 1.11.2 chart, my CRD definitions were out of date. I was able to fix that by completely removing fission with `helm uninstall`, and then manually `kubectl del`eting each of the CRDs, and then installing a fresh release with helm, which created the updated CRDs. I couldn't find where the CRDs were created in the helm chart but I feel like this might be a gotcha for other folks too! 

Thanks!
